webapp/src/components/App/Studies/StudyTree.tsx

```diff
@@ -21,18 +21,127 @@ import TreeItemEnhanced from "../../common/TreeItemEnhanced";
 import { SimpleTreeView } from "@mui/x-tree-view/SimpleTreeView";
 import { getParentPaths } from "../../../utils/pathUtils";
 import * as R from "ramda";
+import { useEffect, useState } from "react";
+import * as api from "../../../services/api/study";
+
+/**
+ * Add a folder that was returned by the explorer into the study tree view.
+ *
+ * This folder isn't a study, otherwise it woudl'nt be
+ * returned by the explorer API, but this folder can have study in it
+ * and still not be in the initial study tree that's parsed from the study
+ * list, this happen when the studies in the folder aren't scanned yet.
+ *
+ * However we want to allow the user to see these folder. When the user explore
+ * they shoudln't wait for a long running scan to complete before they're able to
+ * a folder in the hierarchy.
+ *
+ * @param studiesTree
+ * @param folder
+ * @returns
+ */
+function mergeStudyTreeAndFolder(
+  studiesTree: StudyTreeNode,
+  folder: NonStudyFolder,
+): StudyTreeNode {
+  if (folder.parentPath == studiesTree.path) {
+    for (const child of studiesTree.children) {
+      if (child.name == folder.name) {
+        // parent path is the same, folder name is the same
+        // we don't override the existing folder
+        // so we return here and don't update this node
+        return studiesTree;
+      }
+    }
+    // parent path is the same, but no folder with the same name at this level
+    return {
+      ...studiesTree,
+      children: [
+        ...studiesTree.children,
+        {
+          path: `${folder.parentPath}/${folder.name}`,
+          name: folder.name,
+          children: [],
+        },
+      ],
+    };
+  } else if (folder.parentPath.startsWith(studiesTree.path)) {
+    // recursively walk though the tree
+    // recursively merge in each child
+    return {
+      ...studiesTree,
+      children: studiesTree.children.map((child) =>
+        mergeStudyTreeAndFolder(child, folder),
+      ),
+    };
+  } else {
+    // folder isn't part of this hierarchy
+    return studiesTree;
+  }
+}
+
+/**
+ * Merge several folders in the study tree.
+ *
+ * @param studiesTree
+ * @param folders
+ * @returns
+ */
+export function mergeStudyTreeAndFolders(
+  studiesTree: StudyTreeNode,
+  folders: NonStudyFolder[],
+): StudyTreeNode {
+  let mergedStudyTree = studiesTree;
+  for (const folder of folders) {
+    mergedStudyTree = mergeStudyTreeAndFolder(mergedStudyTree, folder);
+  }
+  return mergedStudyTree;
+}
```

Comment:
suggestion: be careful about state mutations, here you can for example use reduce() method as it handles the fold operation more elegantly while maintaining immutability principles.

```typescript
export const mergeStudyTreeAndFolders = (
  studyTree: StudyTreeNode,
  folders: NonStudyFolder[],
): StudyTreeNode => folders.reduce(
  (tree, folder) => mergeStudyTreeAndFolder(tree, folder), 
  studyTree
);
```